allocator.go

```diff
@@ -184,3 +218,66 @@ func (a *Allocator) Allocate() (ok bool, assignments map[RangeID][]NodeID) {
 	}
 	return true, res
 }
+
+// Minimize the difference absolute difference from the previous assignment and the current assignment
+func (a *Allocator) adhereToMinimalChurnConstraint() {
+
+	if !a.opts.churnOptions.WithMinimalChurn && !a.opts.churnOptions.withMaxChurn {
+		return
+	}
+
+	var vars = make([]cpsatsolver.IntVar, 0, 2*len(a.ranges)*len(a.nodes))
+	var coefficients = make([]int64, 0, len(a.ranges)*len(a.nodes))
+	o := toMap(a.opts.churnOptions.prevAssignment)
+
+	for _, r := range a.ranges {
+		for _, n := range a.nodes {
+			termCoefficient := int64(1)
+			if val := o[r.id][n.id]; val == 1 {
+				prevAllocation := a.model.NewIntVar(1, 1, fmt.Sprintf("Previous: r%d-on-n%d", r.id, n.id))
+				vars = append(vars, prevAllocation)
+				coefficients = append(coefficients, 1)
+				termCoefficient = -1
+			}
+			vars = append(vars, a.literal(r, n))
+			coefficients = append(coefficients, termCoefficient)
+		}
+	}
+
+	churnExpression := cpsatsolver.NewLinearExpr(vars, coefficients, 0)
+	if a.opts.churnOptions.WithMinimalChurn {
+		a.model.Minimize(churnExpression)
+	}
+	if a.opts.churnOptions.withMaxChurn {
+		a.model.AddConstraints(cpsatsolver.NewLinearConstraint(churnExpression, cpsatsolver.NewDomain(math.MinInt64, a.opts.churnOptions.maxChurn)))
+	}
+}
+
+// Minimize the number of nodes which were zero but would now become 1
+func (a *Allocator) adhereToMinimalChurnConstraint2() {
+
+	if !a.opts.churnOptions.WithMinimalChurn && !a.opts.churnOptions.withMaxChurn {
+		return
+	}
+
+	var vars = make([]cpsatsolver.IntVar, 0, len(a.ranges)*len(a.nodes)/2)
+	var coefficients = make([]int64, 0, len(a.ranges)*len(a.nodes)/2)
```

Comment:
Why the `/2`? I think for full correctness this would be `sum over ranges * saidRanges.rf`.

Reply:
`vars` here are only really added to the list if their value was zero, so there is a good chance that not all node/range values are zeroes so I decided to make the capacity half so that even if the capacity is small, the double of it would be the maximum possible.
(Typically if an append occurs to a list that would run out of capacity would copy the list and allocate double the capacity)